In [2]:
import numpy as np
from numpy import linalg as LA
import fractions
from fractions import Fraction
from __future__ import print_function


In [3]:
A = [

    [18, 6, 0], 
    [6, 6, -7],
    [0, -7, 18]
]

A_norm = LA.norm(A, ord = 1)

A_inv = LA.inv(A)

A_inv_norm = LA.norm(A_inv, ord = 1)

print(A_inv_norm)
print("mu = ", A_norm * A_inv_norm)

for a in A_inv: 
    for ai in a: 
        print(Fraction(ai).limit_denominator(max_denominator=100), end = " ")
    print()


1.347826086956522
mu =  33.69565217391305
1/7 -6/23 -7/69 
-6/23 18/23 7/23 
-7/69 7/23 4/23 


## a)

$$||\cdot||_1$$
$$ ||A||_1 = \max\limits_{i} \sum_{j} |a_{ij}|$$
$$ \mu = ||A||_1 \cdot ||A^{-1}||_1 \approx 33.7 $$ 
$$\dfrac{||\Delta x||_1}{||x||_1} \leqslant \mu \dfrac{||\Delta f||_1}{||f||_1} \leqslant 0.34$$

$$ ||\cdot||_2 $$

$$ ||A||_2 = \max\limits_{j} \sum_{i} |a_{ij}|$$

$$ ||A||_2 = ||A||_1 = 25$$
$$ ||A^{-1}||_2 =  ||A^{-1}||_1 = 1.35$$
$$ \dfrac{||\Delta x||_1}{||x||_1} \leqslant 0.34 $$

In [85]:
A_inv2 = LA.matrix_power(A_inv, 2)

print(LA.eig(A_inv2)[0])

[1.         0.00308642 0.00189036]


$\lambda_{(A^{-1})^{*} A^{-1}} \rightarrow \qquad
                                1 , \qquad
                                0.00308 , \qquad
                                0.00189$

$$ ||A^{-1}||_3 = 1 $$
$$\Downarrow$$
$$ \mu = 23 $$
$$\Downarrow$$
$$\dfrac{||\Delta x||_3}{||x||_3} \leqslant 0.23$$


## b)
В оценках $\dfrac{||\Delta x||}{||x||} \leqslant \mu \dfrac{||\Delta f||}{||f||}$ точное равенство достигается в самом худшем случае, когда $\nu(\vec{f}) = \max\nu(\vec{f}) = \mu$

Для $||\cdot||_3$ нужно взять $\vec(f) = \vec(v)_{\max\lambda_i(A)}$

$\max|\lambda_i(A)| = 23$

In [3]:
E = np.eye(3)
print(A - 23*E)

[[ -5.   6.   0.]
 [  6. -17.  -7.]
 [  0.  -7.  -5.]]


$\vec{v} = \begin{pmatrix} -6 \\ -5 \\ 1  \end{pmatrix} = \vec{f} \Rightarrow ||\Delta f ||_3 = 0.01 \cdot \sqrt{36+25+49} \approx 0.1$




$\sum_{i = \overline{1, 3}} \Delta f_i^2 \approx 0.01$

$$ || \cdot ||_1 : $$
$$ \nu(f) = \mu = ||A^{-1}||_1 \cdot || A ||_1 = ||A^{-1}||_1 \cdot \dfrac{||\vec{f}||_1}{||\vec{x}||_1}$$
$$ ||\vec{f}||_1 = || A ||_1 \cdot ||x||_1 =  || A ||_1 \cdot || A^{-1}  \vec{f}||_1$$

## c) 

$\vec{f}$ известен. Найти $\nu(\vec{f})$, которая выполняется $\forall \vec{\Delta f}$

$$ \dfrac{||\Delta x||}{||x||} \leqslant \nu(f) \cdot \dfrac{||\Delta f||}{||f||}  $$

$$ \nu{f} = \dfrac{||f||}{||x||} \cdot ||A^{-1}|| = \dfrac{||f||}{||A^{-1}f||}\cdot ||A^{-1}||$$

$||f||_3 = \sqrt{f_1^2 + f_2^2 + f_3^2}$ \qquad  $ ||A^{-1}||_3 = 1 $ 
$$ ||A^{-1}f||_3 = \dfrac{1}{414} \sqrt{16909f_1^2 + 132516f_2^2 + 22824f_3^2 - 93312f_1 f_2 - 38220f_1f_3 + 108864f_2f_3}$$
$$ \nu(f) = 414 \cdot \dfrac{\sqrt{f_1^2 + f_2^2 + f_3^2}}{\sqrt{16909f_1^2 + 132516f_2^2 + 22824f_3^2 - 93312f_1 f_2 - 38220f_1f_3 + 108864f_2f_3}} $$
$$ \nu(f) \in (1, 23) $$
$$ \max \nu(f) \rightarrow f = (-6, -5, 7)^T  $$
$$ \min \nu(f) \rightarrow f = (-6, 17, 7)^T  $$


## d)
### Метод Якоби

$$ x^{(s+1)} = - D^{-1}(L+U) \cdot x^{(s)}+D^{-1}\cdot f $$

In [20]:
D = np.array([

    [18, 0, 0],
    [0, 6, 0],
    [0, 0, 18]
])

L = np.array([

    [0, 0, 0],
    [6, 0, 0],
    [0, -7, 0]
])

U = np.array([
    [0, 6, 0],
    [0, 0, -7],
    [0, 0, 0]
])

Dinv_LU = np.array([
    [0, 1/3, 0],
    [-1, 0, -7/6],
    [0, -7/18, 0]
])

Dinv = np.array([

    [1/18, 0, 0],
    [0, 1/6, 0],
    [0, 0, 1/18]
])

A = np.array(L + D + U)
print(A)

[[18  6  0]
 [ 6  6 -7]
 [ 0 -7 18]]


Метод Якоби сходится при условии, что все корни уравнения 

$\det \begin{pmatrix} 16 \lambda & 6 & 0\\ 6 & 6\lambda & -7 \\ 0& -7 & 18\lambda  \end{pmatrix} = 0$

по модулю меньше 1.

$|\lambda_1|=0 \qquad |\lambda_2|= 0, 86 \lt 1 \qquad |\lambda_3|= 0, 53 \lt 1$ 

Метод Якоби сходится.

## Метод Зейделя

$$
u^{(k+1)} = R\cdot u^{(k)} + F \\

R = -(L + D)^{-1}U, \qquad F = (L + D )^{-1} f
$$

In [31]:
LD_INV = LA.inv(L+D)*LA.det(L+D)
np.matmul(LD_INV, U)

array([[    0.,   648.,     0.],
       [    0.,  -648., -2268.],
       [    0.,  -252.,  -882.]])

begin{pmatrix} 16 \lambda & 6 & 0\\ 6 & 6\lambda & -7 \\ 0& -7 & 18\lambda  \end{pmatrix} 

In [8]:
3*np.sin(2)**2/(np.sin(2)**2 + 6*np.sin(4))

-0.6678702204044642

In [7]:
-np.log2(10**(-3)/3)

11.550746785383243